reco_utils/recommender/rbm/rbm.py

```diff
@@ -0,0 +1,946 @@
+
+            2) Choose 0 if pr<g, 1 otherwise. It is convenient to implement this condtion using
+               the relu function.
+
+        '''
+
+        np.random.seed(1)
+
+        #sample from a Bernoulli distribution with same dimensions as input distribution
+        g = tf.convert_to_tensor(np.random.uniform(size= pr.shape[1] ), dtype=tf.float32)
+
+        #sample the value of the hidden units
+        h_sampled = tf.nn.relu(tf.sign(pr-g) )
+
+        return h_sampled
+
+    #Multinomial sampling
+    def M_sampling(self, pr):
+
+        '''
+        Multinomial Sampling of ratings
+
+        Args:
+            pr:  a distributions of shape (m, n, r), where m is the number of examples, n the number
+                 of features and r the number of classes. pr needs to be normalized, i.e.
+                 sum_k p(k) = 1 for all m, at fixed n.
+            f:   normalized, uniform probability used for comparison.
+
+        Returns:
+            v_samp: an (m,n) array of sampled rankings from 1 to r .
+
+        Basic mechanics:
+                For r classes, we sample r binomial distributions using the rejection method. This is possible
+                since each class is statistically independent from the other. Note that this is the same method
+                used in numpy's random.multinomial() function.
+
+                1) extract a size r array of random numbers from a uniform distribution (g). As pr is normalized,
+                   we need to normalize g as well.
+
+                2) For each user and item, compare pr with the reference distribution. Note that the latter needs
+                   to be the same for ALL the user/item pairs in the dataset, as by assumptions they are sampled
+                   from a common distribution.
+
+        '''
+
+        np.random.seed(1)
+
+        g = np.random.uniform(size= pr.shape[2]) #sample from a uniform distribution
+        f = tf.convert_to_tensor(g/g.sum(), dtype= tf.float32 )#normalize and convert to tensor
+
+        samp = tf.nn.relu(tf.sign(pr - f) ) #apply rejection method
+        v_samp = tf.cast( tf.argmax(samp , axis=2)+1, 'float32') #select sampled element
+
+        return v_samp
+
+    #Multinomial distribution
+    def Pm(self, phi):
+
+        '''
+        Probability that unit v has value l given phi: P(v=l|phi)
+
+        Args:
+            phi: linear combination of values of the previous layer
+            r  : rating scale, corresponding to the number of classes
+
+        Returns:
+            pr: a tensor of shape (r, m, Nv) . This needs to be reshaped as (m, Nv, r) in the last step
+                to allow for faster sampling when used in the Multinomial function.
+
+        '''
+
+        num = [tf.exp(tf.multiply(tf.constant(k, dtype='float32'),phi)) for k in range(1,self.r_+1)]
+        den = tf.reduce_sum(num, axis=0)
+
+        pr = tf.div(num, den)
+
+        return tf.transpose(pr, perm= [1,2,0])
+
+    #Free energy
+    def Fv(self, x):
```

Comment:
I would recommend to rename these functions to something more descriptive, maybe `free_energy(...)`